content/courses/advanced-ada/parts/resource_management/controlled_types.rst

```diff
@@ -1548,13 +1548,422 @@ Completion
+       --  scope:
+       --
+       --  Finalize (A);
+       --
+    end Show_Finalize_Exception;
+
+In this example, we're again forcing an overflow to happen (by adding one to
+the integer variable :ada:`Cnt`), this time in the :ada:`Finalize` procedure.
+When this procedure is implicitly called |mdash| when object :ada:`A` is about
+to get out of scope in the :ada:`Show_Finalize_Exception` procedure |mdash|
+the :ada:`Constraint_Error` exception is raised.
+
+As we've just seen, having an exception be raised during an implicit call to
+the :ada:`Finalize` procedure is a bounded error. Therefore, we see that the
+:ada:`Program_Error` exception is raised at runtime instead of the original
+:ada:`Constraint_Error` exception.
+
+As we hinted in the beginning, when the :ada:`Adjust` or the :ada:`Finalize`
+procedure is called *explicitly*, the exception raised in that procedure is
+*not* considered a bounded error. In this case, the original exception is
+raised.
+
+To show an example of such an explicit call, let's first move the overriden
+procedures for type :ada:`T` (:ada:`Initialize`, :ada:`Adjust` and
+:ada:`Finalize`) out of the private part of the package
+:ada:`CT_Finalize_Exception`, so they are now visible to clients. This allows
+us to call the :ada:`Finalize` procedure explicitly:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Controlled_Types.Exception_Handling.CT_Finalize_Exception
+
+    with Ada.Finalization;
+
+    package CT_Finalize_Exception is
+
+       type T is new
+         Ada.Finalization.Controlled
+       with null record;
+
+       overriding
+       procedure Initialize (E : in out T);
+
+       overriding
+       procedure Adjust (E : in out T);
+
+       overriding
+       procedure Finalize (E : in out T);
+
+       procedure Dummy (E : T);
+
+       procedure Reset_Counter;
+
+    end CT_Finalize_Exception;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with CT_Finalize_Exception;
+    use  CT_Finalize_Exception;
+
+    procedure Show_Finalize_Exception is
+       A : T;
+    begin
+       Dummy (A);
+
+       Finalize (A);
+
+       Put_Line ("After Finalize");
+    exception
+       when Constraint_Error =>
+          Put_Line ("Constraint_Error is being handled...");
+          Reset_Counter;
+    end Show_Finalize_Exception;
+
+Now, we're calling the :ada:`Finalize` procedure explicitly in the
+:ada:`Show_Finalize_Exception` procedure. As we know, due to the operation on
+:ada:`I` in the :ada:`Finalize` procedure, the :ada:`Constraint_Error`
+exception is raised in the procedure. Because we're handling this exception in
+the :ada:`Show_Finalize_Exception` procedure, we see the corresponding user
+message ("Constraint_Error is being handled...") at runtime.
+
+(Note that in the exception handling block, we're calling the
+:ada:`Reset_Counter` procedure. This prevents the :ada:`Constraint_Error` to be
+raised in the next call to :ada:`Finalize`.)
```

Comment:
Maybe "This prevents the :ada:`Constraint_Error` to be raised in the next call to :ada:`Finalize`." 
should be 
"This prevents :ada:`Constraint_Error` from being raised in the next call to :ada:`Finalize`."

Reply:
Yes, that sounds better.